src/Workspaces/Core/Portable/Execution/AssetBuilder.cs

```diff
@@ -182,13 +187,20 @@ public AssetOnlyTreeNode(Solution solution)
 
             public Serializer Serializer { get; }
 
-            public Task<TResult> GetOrCreateAssetAsync<TKey, TValue, TResult>(
-                TKey key, TValue value, string kind, Func<TValue, string, CancellationToken, Task<TResult>> valueGetterAsync, CancellationToken cancellationToken)
+            public Asset GetOrCreateAsset<TKey, TValue>(
+                TKey key, TValue value, string kind, Func<TValue, string, CancellationToken, Asset> valueGetter, CancellationToken cancellationToken)
                 where TKey : class
-                where TResult : Asset
             {
                 Contract.ThrowIfNull(key);
-                return valueGetterAsync(value, kind, cancellationToken);
+                return valueGetter(value, kind, cancellationToken);
+            }
+
+            public TResult GetOrCreateChecksumObjectWithChildren<TKey, TValue, TResult>(
+                TKey key, TValue value, string kind, Func<TKey, TValue, string, CancellationToken, TResult> valueGetter, CancellationToken cancellationToken)
+                where TKey : class
+                where TResult : ChecksumObjectWithChildren
+            {
+                return Contract.FailWithReturn<TResult>("shouldn't be called");
```

Comment:
Something is very wonkky here with the object hierarchy.  If this is in the interface... then why does this not implmeent it with an actual impl.  A caller shouldn't have to know what actual implementation of the interface they have.  And if they don't know, then this could throw.  If they do know what class they have, then this can go off of the interface and be only on the classes that need it.

--

Basically, this is a big smell.  An indication that different sorts of things are being combined into an object hierarchy that they don't properly fit.